File: Medium/1239.cs
using System;
using System.Collections.Generic;

public class Solution {
  public int MaxLength(IList<string> arr) {
    if (arr == null || arr.Count == 0) {
      return 0;
    }

    // 判断 s 里面的字符是否唯一
    bool IsUniqueChars(string s) {
      HashSet<char> set = new HashSet<char>();

      foreach (char c in s) {
        if (set.Contains(c)) {
          return false;
        }

        set.Add(c);
      }

      return true;
    }

    int result = 0;

    // 组合
    void Dfs(string path, int start) {
      if (IsUniqueChars(path)) {
        result = Math.Max(result, path.Length);
      }
      else {
        return;
      }

      for (int i = start; i < arr.Count; i++) {
        Dfs(path + arr[i], i + 1);
      }
    }

    Dfs("", 0);
    return result;
  }

  public int MaxLength2(IList<string> arr) {
    // 过滤掉包含重复字符的字符串
    // JS: arr = arr.filter(str => new Set(str).size === str.length)
    List<string> filtered = new List<string>();
    foreach (string str in arr) {
      HashSet<char> set = new HashSet<char>(str);

      if (set.Count == str.Length) {
        filtered.Add(str);
      }
    }

    int maxLen = 0;
    HashSet<char> currentStringSet = new HashSet<char>();

    void Backtrack(int index) {
      // 更新最大长度
      maxLen = Math.Max(maxLen, currentStringSet.Count);

      // 遍历剩余字符串
      for (int i = index; i < filtered.Count; i++) {
        // 当前字符串的字符集合
        // 标记是否有冲突
        HashSet<char> strSet = new HashSet<char>(filtered[i]);
        bool hasConflict = false;

        // 检查当前字符串和 currentStringSet 是否有重复字符
        foreach (char ch in strSet) {
          if (currentStringSet.Contains(ch)) {
            hasConflict = true;
            break;
          }
        }

        if (!hasConflict) {
          // 将当前字符串加入组合，并递归处理下一个字符串
          foreach (char ch in strSet) {
            currentStringSet.Add(ch);
          }

          Backtrack(i + 1);

          // 回溯时移除当前字符串的字符
          foreach (char ch in strSet) {
            currentStringSet.Remove(ch);
          }
        }
      }
    }

    Backtrack(0);
    return maxLen;
  }
}

File: Medium/1239.py
class Solution:
    def maxLength(self, arr: List[str]) -> int:
      if len(arr) is 0 or not arr:
        return 0

      # 判断s里面的字符是否唯一
      def isUniqueChars(s: string):
        charSet = set()
        for c in s:
            if c in charSet:
                return False
            charSet.add(c)
        return True

      result = 0
      # 组合
      def dfs(path, start):
        if isUniqueChars(path):
            nonlocal result
            result = max(len(path), result)
        else:
            return

        for i in range(start, len(arr)):
            dfs(path + arr[i], i + 1)

      dfs("", 0)
      return result
    
    from typing import List
    
    def maxLength2(self, arr: List[str]) -> int:
      # 过滤掉包含重复字符的字符串
      # JS: arr = arr.filter(str => new Set(str).size === str.length)
      arr = [s for s in arr if len(set(s)) == len(s)]

      max_len = 0
      current_string_set = set()

      def backtrack(index):
        nonlocal max_len, current_string_set

        # 更新最大长度
        max_len = max(max_len, len(current_string_set))

        # 遍历剩余字符串
        for i in range(index, len(arr)):
          # 当前字符串的字符集合
          # 标记是否有冲突
          str_set = set(arr[i])
          has_conflict = False

          # 检查当前字符串和 current_string_set 是否有重复字符
          for char in str_set:
            if char in current_string_set:
              has_conflict = True
              break

          if not has_conflict:
            # 将当前字符串加入组合，并递归处理下一个字符串
            # JS: currentStringSet = new Set([...currentStringSet, ...strSet])
            current_string_set = current_string_set | str_set

            backtrack(i + 1)

            # 回溯时移除当前字符串的字符
            for char in str_set:
              current_string_set.remove(char)

      backtrack(0)
      return max_len
    
    def maxLength3(self, arr: List[str]) -> int:
      if len(arr) is 0 or not arr:
        return 0

      # 判断s里面的字符是否唯一
      def isUniqueChars(s: string):
        charSet = set()
        for c in s:
            if c in charSet:
                return False
            charSet.add(c)
        return True

      result = 0
      # 组合
      def dfs(path, start):
        if not isUniqueChars(path):
          return;
    
        nonlocal result
        result = max(len(path), result);

        if start >= len(arr): 
          return;
    
        # 选这个单词
        dfs(path + arr[start], start + 1)
        # 不选这个单词
        dfs(path + arr[start], start + 1)
            

      dfs("", 0)
      return result

    